### PR TITLE
fix(cognito): changing `installLatestAwsSdk` breaks Client Secret reference

### PR DESCRIPTION
**Pull request by laurelmay**
Because there wasn't previously a handler for `onUpdate` events, an
empty object would be returned. When `installLatestAwsSdk` was changed
to `false`, this was an update. Typically, updates aren't an issue
because basically any other property being updated signifies a
replacement. `installLatestAwsSdk` is just a very unique case where it
doesn't (and where a user usually can't update it).

When the empty object is returned, this results in an update failure in
CloudFormation because the specific property isn't available.

Fixes: #23796

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Construct Runtime Dependencies:

* [ ] This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-construct-runtime-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:465]: Is there any circumstance where the custom resources wouldn't be in the same region? Probably a dumb question on my end, but I just want to check.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Thanks for the fix! I think this looks good but I just have one question inline.

**Comment by TheRealAmazonKendra** [comment]
Probably shouldn't have actually made that changes requested. Can you tag me when you respond so I see it even if it doesn't require another code change?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (kylelaker) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:465]: @TheRealAmazonKendra That's a great question and honestly I am not entirely sure... So the User Pool and the Client always have to be in the same region. There also isn't (currently) a `fromXXX` static method on `UserPoolClient` that lets you specify region (there's `fromUserPoolClientId` but not `fromUserPoolClientArn`). And while User Pool has a `fromUserPoolArn` and `IUserPool` has `addClient`, I actually feel like that would fail in CloudFormation since the pool is in another region.

And then `fromUserPoolClientId` actually has an implementation of `get userPoolClientSecret()` that always throws (probably because it doesn't take an `IUserPool` as a parameter to make the API call).

I am not sure that there's any situation where the `region` parameter's value could be different from the region where the pool/client exist (and have things actually work with the interfaces that exist today).

So I guess `region` could actually probably be omitted? But frankly, I just copied what had been done for `onCreate`.

Which, I think, means the options are to either:
1. Keep `region` as-is
2. Remove `region` from `onCreate` and `onUpdate`
3. Change both to `this.userPool.env.region` (which at the moment should always be the stack's region)

**Comment by kylelaker** [comment]
So I actually did just push a change... I forgot that when `onCreate` is not defined, `onUpdate` is used instead. So I think we can just replace `onCreate` with `onUpdate`. The region question is still a good one though.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts
Comment (comcalvi) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:454]: Should `onCreate` be completely removed?

Comment (kylelaker) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:454]: Yep! It seems to be relatively unknown that `onCreate` defaults to the value of on `onUpdate` when not specified (see `@default` for https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources.AwsCustomResource.html#oncreate).

Please check me on this and confirm that the changes to integration test snapshots _add_ `Update` but _do not_ remove `Create`?

https://github.com/aws/aws-cdk/blob/4c62f0b31911ededb5542030ee03fec17b3af994/packages/%40aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts#L435-L446

I think this is actually a common bug (specifying `onCreate` but not `onUpdate` for `AwsCustomResource`), even within `@aws-cdk/*`.

Comment (comcalvi) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-client.ts:454]: I did not know that, thanks for the clarification! 

Yep, `OnCreate` is still in the snapshots

**Review by comcalvi — DISMISSED** [review]

**Comment by kylelaker** [comment]
It looks like mergify is upset because the auto-update will result in changing something in `.github/workflows/`. I believe that will require maintainer action anyway, so I am going to go ahead and merge in `main` (which I think will end up dismissing the review...).

**Comment by TheRealAmazonKendra** [comment]
Mergify's gonna mergify... or not. Re-approving this.

**Review by TheRealAmazonKendra — APPROVED** [review]

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**
